oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/AsyncCheckpointCreator.java

```diff
@@ -0,0 +1,143 @@
+ */
+public class AsyncCheckpointCreator implements Runnable {
+
+    /**
+     * Name of service property which determines the name of this Async task
+     */
+    public static final String PROP_ASYNC_NAME = "oak.checkpoint.async";
+
+    private final String name;
+    private final long checkpointLifetimeInSeconds;
+    private final long minConcurrentCheckpoints;
+    private final long maxConcurrentCheckpoints;
+    private final NodeStore store;
+    public static final String CHECKPOINT_CREATOR_KEY = "creator";
+    public static final String CHECKPOINT_CREATED_KEY = "created";
+    public static final String CHECKPOINT_CREATED_TIMESTAMP_KEY= "created-epoch";
+    public static final String CHECKPOINT_THREAD_KEY = "thread";
+    public static final String CHECKPOINT_NAME_KEY = "name";
+    private static final Logger log = LoggerFactory
+            .getLogger(AsyncCheckpointCreator.class);
+
+    public AsyncCheckpointCreator(@NotNull NodeStore store, @NotNull String name, long checkpointLifetimeInSeconds, long minConcurrentCheckpoints, long maxConcurrentCheckpoints) {
+        this.store = store;
+        this.name = name;
+        this.checkpointLifetimeInSeconds = checkpointLifetimeInSeconds;
+        this.minConcurrentCheckpoints = minConcurrentCheckpoints;
+        // maxConcurrentCheckpoints should at least be 1 more than minConcurrentCheckpoints
+        if (maxConcurrentCheckpoints <= minConcurrentCheckpoints) {
+            log.warn("[{}] Max concurrent checkpoints is less than or equal to min concurrent checkpoints. " +
+                    "Setting max concurrent checkpoints to min concurrent checkpoints + 1.", this.name);
+            this.maxConcurrentCheckpoints = minConcurrentCheckpoints + 1;
+        } else {
+            this.maxConcurrentCheckpoints = maxConcurrentCheckpoints;
+        }
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    protected long getCheckpointLifetimeInSeconds() {
+        return checkpointLifetimeInSeconds;
+    }
+
+    protected long getMinConcurrentCheckpoints() {
+        return minConcurrentCheckpoints;
+    }
+
+    protected long getMaxConcurrentCheckpoints() {
+        return maxConcurrentCheckpoints;
+    }
+
+    @Override
+    public void run() {
+        Map<String, Map<String, String>> filteredCheckpointMap = IndexUtils.getFilteredCheckpoints(store,
+                entry -> name.equals(entry.getValue().get(CHECKPOINT_NAME_KEY)));
+        // If the number of checkpoints created by this task are equal to or greater than the maxConcurrentCheckpoints, skip
+        // creation of a new checkpoint.
+        // This could happen in case the deletion of older checkpoints failed in multiple previous runs.
+        if (filteredCheckpointMap.size() >= maxConcurrentCheckpoints) {
+            log.info("[{}] Skipping checkpoint creation as the number of concurrent checkpoints is already at max limit", name);
```

Comment:
I would print this as a WARN
```suggestion
            log.warn("[{}] Skipping checkpoint creation as the number of concurrent checkpoints is already at max limit", name);
```

Reply:
I would also suggest indicating what is the max limit, it can be useful for debugging.